src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.ComposeCache.targets

```diff
@@ -24,25 +24,177 @@ Copyright (c) .NET Foundation. All rights reserved.
+      <Output ItemName="AllResolvedPackagesPublished" TaskParameter="TargetOutputs" />
+    </MSBuild>
+  </Target>
+  <!--
+    ============================================================
+                                        CacheWorkerMapper
+
+   Processes each package specified in a cache project file
+    ============================================================
+    -->
+
+  <Target Name="CacheWorkerMapper"
+          Returns ="@(ResolvedPackagesFromMapper)">
+
+    <ItemGroup>
+      <PackageReferencesToCache Include="$(RestoreProj)">
+        <Properties>
+          CachePackageName=%(PackageReference.Identity);
+          CachePackageVersion=%(PackageReference.Version);
+          ComposeWorkingDir=$(ComposeWorkingDir);
+          PublishDir=$(PublishDir);
+          CacheStagingDir=$(CacheStagingDir);
+          TargetFramework=$(TargetFramework);
+          JitPath=$(JitPath);
+          Crossgen=$(Crossgen);
+          SkipUnchangedFiles=$(SkipUnchangedFiles);
+          PreserveCacheLayout=$(PreserveCacheLayout)
+        </Properties>
+      </PackageReferencesToCache>
+    </ItemGroup>
+
+    <MSBuild Projects="@(PackageReferencesToCache)"
+                 Targets="CacheWorker"
+                 BuildInParallel="$(BuildInParallel)"
+                 ContinueOnError="WarnAndContinue">
+      <Output ItemName="ResolvedPackagesFromMapper" TaskParameter="TargetOutputs" />
+    </MSBuild>
+  </Target>
+
+  <Target Name="CacheWorker"
+          Returns="@(ResolvedPackagesPublished)"
+          DependsOnTargets="PrepforRestoreForComposeCache;
+                            CacheWorkerPerformWork"/>
   
+  <Target Name="CacheWorkerPerformWork"
+          DependsOnTargets="RestoreForComposeCache;
+                            ComputeAndCopyFilesToCacheDirectory;"
+          Condition="!Exists($(CacheWorkerWorkingDir))" />
+
+<!--
+    ============================================================
+                                        CacheFinalizer
+
+   Cleans up and produces artifacts after completion of cache
+    ============================================================
+    -->
+  <UsingTask TaskName="RemoveDuplicatePackageReferences" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
+  <Target Name="CacheFinalizer"
+          DependsOnTargets="CacheWorkerMain;
+                            _CopyResolvedOptimizedFiles">
+
+    <RemoveDuplicatePackageReferences
+         InputPackageRefernces="@(AllResolvedPackagesPublished)">
+      <Output TaskParameter="UniquePackageRefernces"  ItemName="AllResolvedPackagesPublishedAfterFilter"/>
+    </RemoveDuplicatePackageReferences>
+    
+    <ItemGroup>
+      <ListOfPackageReference Include="@(AllResolvedPackagesPublishedAfterFilter -> '%20%20%20%20%20&lt;PackageReference Include=&quot;%(Identity)&quot;  Version =&quot;%(Version)&quot;/&gt;')"/>
+    </ItemGroup>
+    <PropertyGroup>
+     <_CacheArtifactContent>
+      <![CDATA[
+<CacheArtifacts>
+  <ItemGroup>
```

Comment:
As a shepherd of the MSBuild file format: don't use `ItemGroup`. It's a bad name even for MSBuild.